packages/filesystem/README.md

```diff
@@ -0,0 +1,123 @@
+[![stage](https://img.shields.io/endpoint?style=for-the-badge&url=https%3A%2F%2Fraw.githubusercontent.com%2Fsolidjs-community%2Fsolid-primitives%2Fmain%2Fassets%2Fbadges%2Fstage-0.json)](https://github.com/solidjs-community/solid-primitives#contribution-process)
+
+A primitive that allows to manage different file system access methods:
+
+* `createFileSystem` - Provides a reactive interface to one of the file system adapters
+* `makeNoFileSystem` - Adapter that provides a synchronous mock file system
+* `makeNoAsyncFileSystem` - Adapter that provides an asynchronous mock file system
+* `makeVirtualFileSystem` - Adapter that provides a virtual file system that doubles as FsMap for `typescript-vfs` with its `.toMap()` method.
+* `makeWebAccessFileSystem` (client only) - Adapter that provides access to the actual filesystem in the browser using a directory picker
+* `makeNodeFileSystem` (server only) - Adapter that abstracts the node fs/promises module for the use with this primitive
+* `makeTauriFileSystem` (tauri with fs access enabled only) - Adapter that connects to the tauri fs module
+
+## Installation
+
+```bash
+npm install @solid-primitives/filesystem
+# or
+yarn add @solid-primitives/filesystem
+# or
+pnpm add @solid-primitives/filesystem
+```
+
+## How to use it
+
+The synchronous adapters have the following interface:
+
+```tsx
+export type SyncFileSystemAdapter = {
+  async: false;
+  getType: (path: string) => FileType;
+  mkdir: (path: string) => void;
+  readdir: (path: string) => [] | [string, ...string[]];
+  readFile: (path: string) => string;
+  rename?: (previous: string, next: string) => void;
+  rm: (path: string) => void;
+  writeFile: (path: string, data: string) => void;
+};
+```
+
+The asynchronous adapters are returning the same values wrapped in promises:
+
+```ts
+export type AsyncFileSystemAdapter = {
+  async: true;
+  getType: (path: string) => Promise<FileType>;
+  mkdir: (path: string) => Promise<void>;
+  readdir: (path: string) => Promise<[] | [string, ...string[]]>;
+  readFile: (path: string) => Promise<string>;
+  rename?: (previous: string, next: string) => Promise<void>;
+  rm: (path: string) => Promise<void>;
+  writeFile: (path: string, data: string) => Promise<void>;
+};
+```
+
+If no rename method is given, mkdir/writeFile/rm are used to achieve the same effect. An actual rename call might be more performant, though.
+
+The `createFileSystem` call then creates a reactive surface above each of these APIs so that the return values of all reading calls are signals for synchronous and resources for asynchronous filesystem adapters; writing methods for asynchronous APIs will return the same promise for convenience reasons. These getters returned from reading methods are bound to Solid's reactivity so that they will automatically cause effects using them outside of untrack() to re-run on updates. Asynchronous getters will initially return undefined, but will update to the correct value once evaluated.
```

Comment:
A few remarks: 

1. Are changes in "external" filesystems like node propagated into the reactive system? E.g. someone edits a file with node:fs/promises. Is that change reflected in the reactive fs?
2. Why is a sync API supported? Sync calls will block the main thread. Running fs stuff in a worker is complicated. Did someone _require_ a sync API? If not, remove it to reduce the surface area of this primitive?

Reply:
Thanks for your answer.

1. No, we don't receive those changes yet - and even if I did, node:fs/promises is not the only source of possible changes to a file system. I intend to add a `registerWatcher`-method at a later point that allows automatic re-evalutation.
2. The sync API is mainly for virtualized file systems like @typescript/vfs, which is what I need it for; cramming them forcefully into an async interface is bad for the ergonomics - I'd be forced to not rely on Suspense for other operations - but you don't need to use it if you don't want it, this primitive is meant to be fully tree-shakable.